### ISSUE TITLE
etag support

### ISSUE THREAD

**Issue by igorescobar** [dominicbarnes/express-hal#1]
Apparently, express has an built-in e-tag support and it uses it to optimize GET Requests (http://tjholowaychuk.com/post/34189797102/express-3-0). That feature seems to be lost or not functioning as it's supposed to when using the .hal() method instead of .send().

ETag support is extremely important to REST APIs.

**Comment by dominicbarnes** [comment]
It appears that `res.json()` (which I am using under the hood) does _not_ support etags. It would probably be worthwhile to submit this as a feature request with Express directly.

In the interim, I will experiment with using `res.send()` directly (in addition to setting the correct HAL MIME type)

**Comment by dominicbarnes** [comment]
Actually, scratch that, `res.json()` just uses `res.send()`, so it should send an ETag as well. I will investigate this soon. (if you happen across any clues, feel free to post them here)

**Comment by igorescobar** [comment]
That's weird... I had to create/set my ETags manually.

**Comment by igorescobar** [comment]
The .send method take care of the header If-None-Match and If-Match automatically... which means that I don't need to worry about ETags headers. ExpressJs along with .send method does it all. Some how.. It isn't working and it is always returning status 200 no matter what when it should be returning 304.

**Comment by dominicbarnes** [comment]
Are the generated ETag's different with each request? Maybe I'm constructing the response in such a way that the hash that Express calculates changes with each request.

**Comment by igorescobar** [comment]
Nope. Always the same Etag.

**Comment by dominicbarnes** [comment]
I just found [something interesting](https://github.com/visionmedia/express/blob/master/lib/response.js#L136)... Looks like Express will only generate an `ETag` when the `Content-Length` is `1KB` or more.

Also, I tested sending an `If-None-Match` header, and I got the expected `304 Not Modified`. Are you sending the exact header Express sends? (quotes and all?)

**Comment by igorescobar** [comment]
Dude!
Turns out that I found the problem.
I started to debug the "fresh" package that the express uses behind the scenes to deal with etags and cache control headers.

The big shit that I discovered is "POSTMan REST Client (chrome extesion" sends by default the "Cache-Control: no-cache" header behind the scenes without even notify me:

![screen shot 2013-10-15 at 3 04 51 pm](https://f.cloud.github.com/assets/24032/1336218/5b8803d0-35c4-11e3-985f-98e32b24e760.png)

Well... after disabling that disturbing behavior I got the 304.

So sorry disturbing you.
